### DISCUSSION TITLE
Packaging zrepl for illumos

### DISCUSSION THREAD

**Discussion by jgmills** [zrepl/zrepl#617] [Q&A] [answered]
The procedure for building an illumos package consists of a series of steps (or stages or targets).  The ones I normally use in development are: download, unpack, patch, build, install, publish.  The download step, for example, is supposed to download the source archive into a cache.  The unpack step is supposed to unpack the archive into files within the component directory.  This procedure works well for most products available on the Internet, but not for zrepl or indeed any GO products.

During the build step for zrepl, it downloads 46 GO modules to my GOPATH location.  Is there a way to prevent this from happening?  Is there a way to move this operation out of the build step, so that I could do it alone in an earlier step?  Even better, could I obtain these modules in an archive file that could be downloaded in the download step?

**Comment by problame** [comment, marked as answer]
zrepl uses Go modules, so, I suggest you familiarize yourself with that first:  https://go.dev/ref/mod

I guess searching for "vendoring Go modules" should give you what you're looking for.

But I have no interest in making the vendoring part of the zrepl build / release process.

Also, you might want to look into the pain that Debian goes through with packaging Go libraries: https://go-team.pages.debian.net/packaging.html

**Reply by jgmills** [reply]
Thanks for the response.  The first URL was quite useful.  However, vendoring is not quite what I wanted.  I'd like to split the download portion off from the build process so that I could move it an earlier target in the OI Makefile.  As far as I can tell, this is not possible.

I found the second URL to be not useful:  It is mostly policy, rather than technical details.

I've asked on an OI (illumos) mailing list if there is any point it moving the download portion.  So far, I have not received a reply.
